pkg/storage/s3.go

```diff
@@ -1,56 +1,229 @@
+	secretAccessKeyEnv   = "AWS_SECRET_ACCESS_KEY"
+	// number of retries to make of operations
+	maxRetries = 3
 )
 
-// S3BackendOptions are options for configurating the S3 storage.
+// S3BackendOptions contains options for s3 storage
 type S3BackendOptions struct {
-	Endpoint string `json:"endpoint" toml:"endpoint"`
-	Region   string `json:"region" toml:"region"`
+	Endpoint        string `json:"endpoint" toml:"endpoint"`
+	Region          string `json:"region" toml:"region"`
+	StorageClass    string `json:"storage_class" toml:"storage_class"`
+	SSE             string `json:"sse" toml:"sse"`
+	ACL             string `json:"acl" toml:"acl"`
+	ForcePathStyle  bool   `json:"force_path_style" toml:"force_path_style"`
+	AccessKey       string `json:"access_key" toml:"access_key"`
+	SecretAccessKey string `json:"secret_access_key" toml:"secret_access_key"`
 }
 
 func (options *S3BackendOptions) apply(s3 *backup.S3) error {
 	if options.Endpoint == "" && options.Region == "" {
 		return errors.New("must provide either 's3.region' or 's3.endpoint'")
 	}
-	// TODO: Verify Region.
+	if options.AccessKey == "" && options.SecretAccessKey != "" {
+		return errors.New("secret_access_key not found")
+	}
+	if options.AccessKey != "" && options.SecretAccessKey == "" {
+		return errors.New("access_key not found")
+	}
 
+	// StorageClass, SSE and ACL are acceptable to be empty
 	s3.Endpoint = options.Endpoint
 	s3.Region = options.Region
+	s3.StorageClass = options.StorageClass
+	s3.Sse = options.SSE
+	s3.Acl = options.ACL
+	s3.AccessKey = options.AccessKey
+	s3.SecretAccessKey = options.SecretAccessKey
+	s3.ForcePathStyle = options.ForcePathStyle
 	return nil
 }
 
 func defineS3Flags(flags *pflag.FlagSet) {
-	flags.String(s3EndpointOption, "", "Set the AWS S3 endpoint URL")
-	flags.String(s3RegionOption, "", "Set the AWS region")
-	// TODO: Finalize the list of options.
-	_ = flags.MarkHidden(s3EndpointOption)
+	flags.String(s3EndpointOption, "", "Set the S3 endpoint URL")
+	flags.String(s3RegionOption, "", "Set the S3 region")
+	flags.String(s3StorageClassOption, "", "Set the S3 storage class")
+	flags.String(s3SSEOption, "", "Set the S3 server-side encryption algorithm")
+	flags.String(s3ACLOption, "", "Set the S3 canned ACLs")
+	flags.String(s3ProviderOption, "", "Set the S3 provider")
 }
 
 func getBackendOptionsFromS3Flags(flags *pflag.FlagSet) (options S3BackendOptions, err error) {
+	send, err := flags.GetBool(flagSendCredentialOption)
+	if err != nil {
+		err = errors.Trace(err)
+		return
+	}
+	if send {
+		options.AccessKey = os.Getenv(accessKeyEnv)
+		options.SecretAccessKey = os.Getenv(secretAccessKeyEnv)
+	}
 	options.Endpoint, err = flags.GetString(s3EndpointOption)
 	if err != nil {
 		err = errors.Trace(err)
 		return
 	}
-
+	if options.Endpoint != "" {
+		if !strings.HasPrefix(options.Endpoint, "https://") &&
+			!strings.HasPrefix(options.Endpoint, "http://") {
+			options.Endpoint = "http://" + options.Endpoint
+		}
+	}
```

Comment:
Please put this fixup into `apply()`. So that if we get `S3BackendOptions` from a config file (TOML/JSON), we still have the same fix.

(This also means perhaps Provider should be part of `S3BackendOptions` and the Getenv stuff should appear in `apply()`.)

Reply:
done